Libraries/Opc.Ua.Client/Session/Session.cs

```diff
@@ -3652,6 +3745,279 @@ public bool RemoveTransferredSubscription(Subscription subscription)
+                do
+                {
+                    int badNoCPErrorsPerPass = 0;
+                    int badCPInvalidErrorsPerPass = 0;
+                    int otherErrorsPerPass = 0;
+                    uint maxNodesPerBrowse = OperationLimits.MaxNodesPerBrowse;
+
+                    if (excecuteDefensively && ServerMaxContinuationPointsPerBrowse > 0)
+                    {
+                        maxNodesPerBrowse = ServerMaxContinuationPointsPerBrowse < maxNodesPerBrowse ? ServerMaxContinuationPointsPerBrowse : maxNodesPerBrowse;
+                    }
+
+                    // split input into batches
+                    int batchCount = 0;                    
+
+                    List<NodeId> nodesToBrowseForNextPass = new List<NodeId>();
+                    List<ReferenceDescriptionCollection> referenceDescriptionsForNextPass = new List<ReferenceDescriptionCollection>();
+                    List<ServiceResult> errorsForNextPass = new List<ServiceResult>();
+
+                    // loop over the batches
+                    foreach ( var nodesToBrowseBatch in ( (List<NodeId>)nodesToBrowseForPass ).Batch<NodeId, List<NodeId>>( maxNodesPerBrowse ))
+                    {
+                        int batchOffset = batchCount * (int)maxNodesPerBrowse;
+
+                        BrowseWithBrowseNext(
+                                view,
+                                nodesToBrowseBatch,
+                                maxResultsToReturn,
+                                browseDirection,
+                                referenceTypeId,
+                                includeSubtypes,
+                                nodeClassMask,
+                                out var resultForBatch,
+                                out var errorsForBatch
+                            );
+
+                        for (int ii = 0; ii < nodesToBrowseBatch.Count; ii++)
+                        {
+                            if(errorsForBatch[ii].StatusCode == StatusCodes.BadNoContinuationPoints ||
+                                errorsForBatch[ii].StatusCode == StatusCodes.BadContinuationPointInvalid)
+                            {
+                                nodesToBrowseForNextPass.Add( nodesToBrowseForPass[batchOffset + ii] );                                
+                                referenceDescriptionsForNextPass.Add( resultForPass[batchOffset + ii] );                                
+                                errorsForNextPass.Add( errorsForPass[batchOffset + ii] );
+                            }
+
+                            resultForPass[batchOffset + ii].Clear();
+                            resultForPass[batchOffset + ii].AddRange( resultForBatch[ii] );
+                            errorsForPass[batchOffset + ii] = errorsForBatch[ii];
+                            
+                        }
+                        int badNoCp = errorsForBatch.Count(x => x.StatusCode == StatusCodes.BadNoContinuationPoints);
```

Comment:
more efficient would be to do the counting in the upper loop.

Reply:
yes, agreed. Will be changed with the next commit.